### PR TITLE
address websocket review

### PR DESCRIPTION
**Pull request by AviAvni**
## Summary by CodeRabbit

- **New Features**
  - Enhanced the character limit for certain inputs to improve user experience.
  - Implemented version checking during the handshake process for better compatibility management.

- **Bug Fixes**
  - Improved buffer handling and WebSocket communication to ensure more reliable data transfer.

- **Refactor**
  - Streamlined WebSocket frame header writing into a dedicated function for increased code clarity.

- **Documentation**
  - Updated internal documentation to reflect changes in buffer handling functions.

- **Tests**
  - Expanded test coverage with new parameters to validate environment setup.

### CODE REVIEW

File: src/bolt/bolt_client.c [hunk 1]
```diff
@@ -35,6 +35,7 @@ bolt_client_t *bolt_client_new
 	buffer_new(&client->msg_buf);
 	buffer_new(&client->read_buf);
 	buffer_new(&client->write_buf);
+	buffer_index(&client->read_buf, &client->ws_frame, 0);
```
Comment (swilly22) [inline src/bolt/bolt_client.c:38]: I think renaming `buffer_index` to `buffer_set_index` would be a bit clearer.
Also consider placing the modified argument (`client->ws_frame`) as the first argument to the function.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/bolt/ws.c [hunk 2]
```diff
@@ -6,33 +6,111 @@
...
+		return false;
 	}
-	return false;
+	if(headers->sec_ws_version != 13) {
+		return false;
+	}
+	return true;
+}
+
+static void free_headers
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/bolt/ws.c:101]: yes

File: src/bolt/buffer.h [hunk 3]
```diff
@@ -137,6 +137,14 @@ void buffer_write
 	uint32_t size         // size
 );
 
+// apply the mask to the buffer
+void buffer_apply_mask
+(
+	buffer_index_t buf,    // buffer
+	uint32_t masking_key,  // masking key
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/bolt/buffer.h:144]: yes

File: src/bolt/buffer.c [hunk 4]
```diff
@@ -314,6 +314,45 @@ void buffer_write
...
+	char *payload = buf.buf->chunks[buf.chunk] + buf.offset;
+	int local_offset = *offset;
+	uint32_t double_mask[2] = {masking_key, masking_key};
+	uint32_t offset_mask = *(uint32_t *)((char *)double_mask + local_offset);
+	int i = 0;
+	for(; i + 4 <= payload_len; i+=4) {
+		*(uint32_t *)(payload + i) ^= offset_mask;
+	}
+	for(; i < payload_len; i++) {
+		payload[i] ^= ((char*)double_mask)[local_offset++];
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/bolt/buffer.c:344]: add 8 bytes step

File: src/bolt/ws.c [hunk 5]
```diff
@@ -15,24 +17,83 @@
...
+	}
+	while(buffer_index_length(request) > 0) {
+		char *field = buffer_index_read_until(request, ':');
+		if(field == NULL) {
+			break;
+		}
+		buffer_index_read(request, NULL, 2);
+		char *value = buffer_index_read_until(request, '\r');
+		if(value == NULL) {
+			break;
```
Comment (swilly22) [inline src/bolt/ws.c:45]: breaking here would leak `field`

File: src/bolt/ws.c [hunk 6]
```diff
@@ -15,24 +17,83 @@
...
+		}
+		buffer_index_read(request, NULL, 2);
```
Comment (swilly22) [inline src/bolt/ws.c:47]: `buffer_index_read` I would consider introducing a `buffer_index_skip` or `buffer_index_advance` method, which internally will call `buffer_index_read`

File: src/bolt/ws.c [hunk 7]
```diff
@@ -15,24 +17,83 @@
...
+		raxInsert(headers, (unsigned char *)field, strlen(field), (void *)rm_strdup(value), NULL);
```
Comment (swilly22) [inline src/bolt/ws.c:48]: Why do we need to duplicate `value` ?

File: src/bolt/ws.c [hunk 8]
```diff
@@ -45,15 +106,16 @@ bool ws_handshake
 	ASSERT(request != NULL);
 	ASSERT(response != NULL);
 
-	char *sec_ws_key;
-	if(!parse_headers(request, &sec_ws_key)) {
+	rax *headers = raxNew();
+	if(!parse_headers(request, headers) || !validate_headers(headers)) {
 		return false;
 	}
```
Comment (swilly22) [inline src/bolt/ws.c:112]: might be leaking `headers`

File: src/bolt/buffer.c [hunk 9]
```diff
@@ -314,6 +360,48 @@ void buffer_write
...
+) {
+	char *payload = buf.buf->chunks[buf.chunk] + buf.offset;
+	int local_offset = *offset;
+	uint32_t double_mask[4] = {masking_key, masking_key, masking_key, masking_key};
+	uint64_t offset_mask = *(uint64_t *)((char *)double_mask + local_offset);
+	int i = 0;
+	for(; i + 8 <= payload_len; i+=8) {
+		*(uint64_t *)(payload + i) ^= offset_mask;
+	}
+	for(; i < payload_len; i++) {
```
Comment (swilly22) [inline src/bolt/buffer.c:379]: Why don't we go from 8 to 4 to 1 ?

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/bolt/buffer.c (hunk 9 above)
Comment (AviAvni) [reply src/bolt/buffer.c:379]: after we finish with 8 the would be max of 1 4 bytes so don't think is is worth it

**Review by swilly22 — APPROVED** [review]

**Pull request merged**